### PR TITLE
Retry the setup of an ingestion reader when it fails

### PR DESCRIPTION
**Pull request by benzekrimaha**
When a reader's setup fails (typically the getRaftId call to the source), it was dropped from the pending list and never set up again: the reader stays registered in _ingestionSources, so applyUpdates considers the bucket already handled and only refreshes it. The bucket was therefore left out of ingestion until the pod was restarted.

Queue the reader again instead, so it is retried on the next cycle. This follows what _processLogReaderEntries already does with read failures, and keeps the other locations ingesting, which restarting the pod on a failed healthcheck would not.

The reader is only queued again if its source is still configured, as it may have been removed while its setup was in progress.

Issue: BB-846

### CODE REVIEW

File: lib/queuePopulator/IngestionPopulator.js [hunk 1]
```diff
@@ -618,12 +618,22 @@ class IngestionPopulator {
...
+                // sources that are already registered, so a reader dropped
+                // here would never be set up again.
+                const zenkoBucket = logReader.getTargetZenkoBucketName();
+                this.log.error('error setting up log reader, retrying later', {
                     method: 'IngestionPopulator._setupUpdatedReaders',
+                    zenkoBucket,
                     error: err,
                 });
+                // the source may have been removed from the configuration
+                // while its setup was in progress
```
Comment (francoisferrand) [inline lib/queuePopulator/IngestionPopulator.js]: this comment is misleading: it is not that it "may have been removed", it is certain that is was removed c.f. the beginning of the function:

```
        const newReaders = this.logReadersUpdate;
        this.logReadersUpdate = [];
```

→ here you want to explain why you re-add it
```suggestion
                // Setup failures are usually transient (source unreachable,
                // invalid credentials...): queue the reader again so that it
                // is retried on the next cycle
```

File: lib/queuePopulator/IngestionPopulator.js [hunk 2]
```diff
@@ -618,12 +618,22 @@ class IngestionPopulator {
...
         async.each(newReaders, (logReader, cb) => logReader.setup(err => {
             if (err) {
-                // if setup fails for a log reader, don't add it to `logReaders`
-                // log the error and continue setting up others
-                this.log.fatal('error setting up log reader', {
+                // Setup failures are usually transient (source unreachable,
+                // invalid credentials...). Queue the reader again so that it
+                // is retried on the next cycle: `applyUpdates` only refreshes
+                // sources that are already registered, so a reader dropped
+                // here would never be set up again.
```
Comment (francoisferrand) [inline lib/queuePopulator/IngestionPopulator.js]: not relevant to the code: the old comment was -and still is- correct, we don't add this one to logReaders and continue setting up others...

```suggestion
                // if setup fails for a log reader, don't add it to `logReaders`
                // log the error and continue setting up others
```

File: lib/queuePopulator/IngestionPopulator.js [hunk 3]
```diff
@@ -618,12 +618,22 @@ class IngestionPopulator {
...
+                const zenkoBucket = logReader.getTargetZenkoBucketName();
+                this.log.error('error setting up log reader, retrying later', {
                     method: 'IngestionPopulator._setupUpdatedReaders',
+                    zenkoBucket,
```
Comment (francoisferrand) [inline lib/queuePopulator/IngestionPopulator.js]: I think we should log the _location_ instead of the bucket, seems more appropriate - or in addition to the bucket, for consistency with other logs in this file:

```suggestion
                this.log.error('error setting up log reader, retrying later', {
                    method: 'IngestionPopulator._setupUpdatedReaders',)
                    zenkoBucket: logReader.getTargetZenkoBucketName(),
                    location: logReader.getLocationConstraint(),
```

**Comment by francoisferrand** [comment]
not the right target branch: no point fixing so far back in time: from 9.5 is enough
(also this is not an improvement, but a bug. so branch name is not right)

File: lib/queuePopulator/IngestionPopulator.js (hunk 2 above)
Comment (benzekrimaha) [reply lib/queuePopulator/IngestionPopulator.js]: Restored.

File: lib/queuePopulator/IngestionPopulator.js (hunk 3 above)
Comment (benzekrimaha) [reply lib/queuePopulator/IngestionPopulator.js]: Done, `location` sits next to `zenkoBucket` now. I kept the `zenkoBucket` const since the guard right below reads it.

File: lib/queuePopulator/IngestionPopulator.js (hunk 1 above)
Comment (benzekrimaha) [reply lib/queuePopulator/IngestionPopulator.js]: Reworded to say why the reader is queued again.

One precision : that comment was about `_ingestionSources`, the registry of configured sources, not about `logReadersUpdate` which the function drains at the top. Those are two different structures, and the guard reads the former.
It was imprecise for another reason though. `_closeLogState` only deletes the `_ingestionSources` entry when it finds the reader in `logReadersUpdate` or in `logReaders`, and while its setup is in progress the reader is in neither, so a source removed right then still leaves its entry behind. What the guard actually catches is a reader that is no longer the one registered for its bucket, which is what the new comment says.

**Comment by benzekrimaha** [comment]
Retargeted to `development/9.5` and rebased on it, so the diverged-branch warning is gone.

On the branch prefix: BB-846 is already typed Bug in Jira, only the branch name says otherwise. Renaming it means closing this PR and opening a new one, which also moves the base of #2821 stacked on top, so I left the branch name as is.

**Comment by francoisferrand** [comment]
What happens when the location is deleted?

Please confirm there no risk of path or race condition where we would keep retrying to setup a location which was removed...

**Comment by benzekrimaha** [comment]
Fixed here as suggested, in a second commit.
`_closeLogState` now unregisters the source unconditionally, which makes `_ingestionSources` authoritative, and the guard moved above the error branch so it covers the success path too.
The point about the extra ingestion round is addressed rather than just bounded: the source is unregistered as soon as it is removed, so a reader whose setup was in flight is neither queued again nor activated. The extra round does not happen at all.
Additionnal validation made by reverting the fix while keeping the tests gives 6 failures and 24 passes, and the 6 failures are exactly the new cases, the pre-existing behaviours pass either way.
One detail worth flagging for future readers: the active-reader test pins that `_removeReaderState` receives the reader object. The reader has to be read before its entry is deleted, since `_removeReaderState` polls it on a timer until its batch completes, so a later simplification that drops that ordering would throw every 2s instead of failing loudly.

**Review by delthas — APPROVED** [review]
LGTM high level, but I'm not very familiar with that part of the code, so needs a proper second review 😛

File: tests/unit/ingestion/IngestionPopulator.js [hunk 4]
```diff
@@ -278,6 +278,322 @@ describe('Ingestion Populator', () => {
...
+            return logReader;
+        }
+
+        beforeEach(() => {
+            ip.logReaders = [];
+            ip.logReadersUpdate = [];
+        });
+
+        it('should activate a log reader once its setup succeeds', done => {
+            const logReaderMock = createLogReaderMock(WORKING_BUCKET, null);
```
Comment (francoisferrand) [inline tests/unit/ingestion/IngestionPopulator.js]: nit: calling `createLogReaderMock()` is not very readable, the usual pattern is to setup a logReader in the fixture (`beforeEach`),and let each test setup expectations (`logReader.setup.yieldsAsync(setupError)`)

Of you stick with this form to reduce dedup, than should the next 2 lines as well into that helper

```js
function createLogReaderMock(zenkoBucket, setupError) {
	const logReader = sinon.createStubInstance(IngestionReader);
	logReader.getTargetZenkoBucketName.returns(zenkoBucket);
	logReader.setup.yieldsAsync(setupError);

	ip._ingestionSources[zenkoBucket] = logReader;
	ip.logReadersUpdate = [logReader];

	return logReader;
}
```

(mixing both approaches makes it a bit weird)

File: tests/unit/ingestion/IngestionPopulator.js [hunk 5]
```diff
@@ -278,6 +278,322 @@ describe('Ingestion Populator', () => {
...
+            ip._setupUpdatedReaders(err => {
+                assert.ifError(err);
+                assert.deepStrictEqual(ip.logReaders, [logReaderMock]);
+                assert.deepStrictEqual(ip.logReadersUpdate, []);
+                done();
+            });
+        });
+
+        it('should queue a log reader again when its setup fails', done => {
+            const logReaderMock =
```
Comment (francoisferrand) [inline tests/unit/ingestion/IngestionPopulator.js]: line break not needed I guess, fits within max column in our prettier config?
(same in following tests)

File: tests/unit/ingestion/IngestionPopulator.js [hunk 6]
```diff
@@ -278,6 +278,322 @@ describe('Ingestion Populator', () => {
...
+            ip._setupUpdatedReaders(err => {
+                assert.ifError(err);
+                assert.deepStrictEqual(ip.logReaders, [workingReader]);
+                assert.deepStrictEqual(ip.logReadersUpdate, [failingReader]);
+                done();
+            });
+        });
+    });
+
+    describe('removing a source while its setup is in flight', () => {
```
Comment (francoisferrand) [inline tests/unit/ingestion/IngestionPopulator.js]: this describe section is weird: this is still testing `_setupUpdatedReaders()`

File: lib/queuePopulator/IngestionPopulator.js [hunk 7]
```diff
@@ -613,13 +613,26 @@ class IngestionPopulator {
         const newReaders = this.logReadersUpdate;
         this.logReadersUpdate = [];
         async.each(newReaders, (logReader, cb) => logReader.setup(err => {
+            const zenkoBucket = logReader.getTargetZenkoBucketName();
+            // The reader is in neither list while its setup is in flight, so
+            // its source may have been removed or replaced in the meantime. It
+            // must then neither be activated nor retried.
+            if (this._ingestionSources[zenkoBucket] !== logReader) {
+                return cb();
```
Comment (francoisferrand) [inline lib/queuePopulator/IngestionPopulator.js:629]: should add a log message (just info is enough), would help to troubleshoot if something ever happens in this corner case...

File: tests/unit/ingestion/IngestionPopulator.js [hunk 8]
```diff
@@ -278,6 +278,322 @@ describe('Ingestion Populator', () => {
         });
     });
 
+    describe('_setupUpdatedReaders', () => {
+        const FAILING_BUCKET = 'failing-zenko-bucket';
+        const WORKING_BUCKET = 'working-zenko-bucket';
```
Comment (francoisferrand) [inline tests/unit/ingestion/IngestionPopulator.js]: nit: nothing working or failing for these buckets. Only ingestion is failing, not the buckets

putting the name here (which may be needed because the mocks are hidden deep down c.f. https://github.com/scality/backbeat/pull/2806/changes#r3880675313) gives a false impression there is something, where in fact it is just the ingestion setup mock which fail...

File: tests/unit/ingestion/IngestionPopulator.js (hunk 4 above)
Comment (francoisferrand) [reply tests/unit/ingestion/IngestionPopulator.js]: or even just just use the helper to create/add the reader ; but keep the useful mock call (`setup`) in the test

```js
function addLogReader(zenkoBucket) {
	const logReader = sinon.createStubInstance(IngestionReader);
	logReader.getTargetZenkoBucketName.returns(zenkoBucket);

	ip._ingestionSources[zenkoBucket] = logReader;
	ip.logReadersUpdate = [logReader];

	return logReader;
}

it(....) {
   addLogReader("bucket1").setup.yieldsAsync(errors.InternalError);
}
```

In addition to making the mocks more explicit, it would also make the tests more readable, by removing extraneous list modifications - like the following, where the test needs to manually edit only to simulate a race condition (nothing needed for simple case):
```js
// should not queue a log reader again when its setup fails and its source is no longer configured
createLogReaderMock("bucket1")
delete ip._ingestionSources["bucket1"]; // remove from the config
```

File: tests/unit/ingestion/IngestionPopulator.js (hunk 6 above)
Comment (francoisferrand) [reply tests/unit/ingestion/IngestionPopulator.js]: it tests the interaction between _setupUpdatedReaders and _closeLogState, but is it not redundant?

e.g. if you validate on one side the way `_setupUpdatedReaders` behaves (vs the various contents of `_ingestionSources` /  `logReadersUpdate`) and the way _closeLogState updates these : then there is nothing left.... if however you want a more "integrated" test, should use the other methods to setup the readers (like `applyUpdates` and/or `addNewLogSource`)

**Review by francoisferrand — APPROVED** [review]

**Comment by francoisferrand** [comment]
should this land on `9.5` or just delay until `9.6` ?

File: tests/unit/ingestion/IngestionPopulator.js (hunk 6 above)
Comment (benzekrimaha) [reply tests/unit/ingestion/IngestionPopulator.js]: Dropped that describe. `_setupUpdatedReaders` already covers missing/replaced `_ingestionSources`, and `_closeLogState` already covers unregistration including in-flight setup.

File: tests/unit/ingestion/IngestionPopulator.js (hunk 8 above)
Comment (benzekrimaha) [reply tests/unit/ingestion/IngestionPopulator.js]: Renamed to `bucket1` / `bucket2`.

File: tests/unit/ingestion/IngestionPopulator.js (hunk 4 above)
Comment (benzekrimaha) [reply tests/unit/ingestion/IngestionPopulator.js]: Switched to `addLogReader` and left `setup.yieldsAsync(...)` in each test.

File: lib/queuePopulator/IngestionPopulator.js (hunk 7 above)
Comment (benzekrimaha) [reply lib/queuePopulator/IngestionPopulator.js:629]: Added an info log with `zenkoBucket` and `location`.

File: tests/unit/ingestion/IngestionPopulator.js (hunk 5 above)
Comment (benzekrimaha) [reply tests/unit/ingestion/IngestionPopulator.js]: Gone with the helper.

File: lib/queuePopulator/IngestionPopulator.js [hunk 9]
```diff
@@ -613,13 +613,32 @@ class IngestionPopulator {
         const newReaders = this.logReadersUpdate;
         this.logReadersUpdate = [];
         async.each(newReaders, (logReader, cb) => logReader.setup(err => {
+            const zenkoBucket = logReader.getTargetZenkoBucketName();
+            // The reader is in neither list while its setup is in flight, so
+            // its source may have been removed or replaced in the meantime. It
+            // must then neither be activated nor retried.
+            if (this._ingestionSources[zenkoBucket] !== logReader) {
+                this.log.info('skipping log reader after setup: source was ' +
+                    'removed or replaced', {
```
Comment (francoisferrand) [inline lib/queuePopulator/IngestionPopulator.js]: line break not needed (esp in the middle of the string)

**Review by francoisferrand — APPROVED** [review]
